golang/cosmos/x/swingset/keeper/extension_snapshotter.go

```diff
@@ -0,0 +1,270 @@
+package keeper
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/x/swingset/types"
+	vstoragetypes "github.com/Agoric/agoric-sdk/golang/cosmos/x/vstorage/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	snapshots "github.com/cosmos/cosmos-sdk/snapshots/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+var _ snapshots.ExtensionSnapshotter = &SwingsetSnapshotter{}
+
+// SnapshotFormat 1 is a proto message containing an artifact name, and the binary artifact data
+const SnapshotFormat = 1
+
+type activeExtensionSnapshot struct {
+	// The block height of the snapshot in progress
+	height uint64
+	// The logger for this snapshot
+	logger log.Logger
+	// The function to retrieve the exported artifacts
+	retrieveSwingStoreExport func() error
+	// The writer for extensions payloads
+	payloadWriter snapshots.ExtensionPayloadWriter
+}
+
+// SwingsetSnapshotter manages Swingset snapshots, ensuring insensitivity to
+// sub-block timing of communication with the JS side, and enforcing
+// concurrency requirements.
+// The caller of this submodule must arrange block level commit synchronization,
+// to ensure the results are deterministic.
+type SwingsetSnapshotter struct {
+	isConfigured            func() bool
+	takeAppSnapshot         func(height int64)
+	newRestoreContext       func(height int64) sdk.Context
+	swingStoreSnapshotter   *SwingStoreSnapshotter
+	getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry
+	logger                  log.Logger
+	activeExtensionSnapshot *activeExtensionSnapshot
+}
+
+// NewSwingsetSnapshotter creates a new SwingsetSnapshotter
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, swingStoreSnapshotter *SwingStoreSnapshotter, getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry) SwingsetSnapshotter {
```

Comment:
The normal Go convention is for a NewFoo function to return *Foo (i.e., a _pointer_ to a Foo rather than a Foo itself). I don't know how rigid the convention is, so we may decide that this is fine, but on the other hand it's rather low effort to conform with it.

Likewise for the other New… functions.

Reply:
I'm still very confused in golang as to what needs to be a pointer, and what can stay a struct value, and our codebase doesn't help clarify this confusion.